openwillis/measures/video/eye_blink.py

```diff
@@ -0,0 +1,145 @@
+    """
+
+    A = dist.euclidean(eye[1], eye[5])
+    B = dist.euclidean(eye[2], eye[4])
+    C = dist.euclidean(eye[0], eye[3])
+
+    ear = (A + B) / (2.0 * C)
+
+    return ear
+
+def eye_blink_counter(video_directory, device='laptop'):
+    """
+    ---------------------------------------------------------------------------------------------------
+
+    This function counts the number of eye blinks in a given video.
+
+    Parameters:
+    ............
+    video_directory : string
+        The directory of the video to be analyzed
+    device : string
+        The device used to record the video. It can be either 'laptop' or 'mobile'
+
+    Returns:
+    ............
+    framewise: np.ndarray
+        An array containing the frame number and the eye aspect ratio (EAR) of each frame
+    blinks : np.ndarray
+        An array containing for each blink the frame number and the time (in seconds)
+         start of the blink and end of the blink
+    summary : list
+        The number of eye blinks and blink rate (blinks per minute)
+
+    Raises:
+    ............
+    ValueError
+        If device is not 'laptop' or 'mobile'
+
+    ---------------------------------------------------------------------------------------------------
+    """
+
+    if device == 'laptop':
+        prominence, width = 0.1, 3
+    elif device == 'mobile':
+        prominence, width = 0.02, 0.1
+    else:
+        raise ValueError("device can be either 'laptop' or 'mobile'")
+
+    # Initialize MediaPipe Face Mesh
+    print("[INFO] initializing MediaPipe Face Mesh...")
+    mp_face_mesh = mp.solutions.face_mesh
+    face_mesh = mp_face_mesh.FaceMesh()
+
+    print("[INFO] starting video stream thread...")
+    vs = cv2.VideoCapture(video_directory)
+    fps = vs.get(cv2.CAP_PROP_FPS)
+    time.sleep(1.0)
+
+    framewise = []
+    blinks = []
+    frame_n = 0
+
+    while True:
+        ret, frame = vs.read()
+        if not ret:
+            break
+
+        frame_n += 1
+        frame = cv2.resize(frame, (450, int(frame.shape[0] * (450. / frame.shape[1]))))
+
+        # Convert the BGR image to RGB and process it with MediaPipe Face Mesh.
+        results = face_mesh.process(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
+
+        if results.multi_face_landmarks:
+            face_landmarks = results.multi_face_landmarks[0]
+
+            leftEye = np.array([(face_landmarks.landmark[lidx].x, face_landmarks.landmark[lidx].y) for lidx in LEFT_EYE_INDICES], dtype=np.float32)
+            rightEye = np.array([(face_landmarks.landmark[ridx].x, face_landmarks.landmark[ridx].y) for ridx in RIGHT_EYE_INDICES], dtype=np.float32)
+
+            # Calculate the EAR for each eye
+            leftEAR = eye_aspect_ratio(leftEye)
+            rightEAR = eye_aspect_ratio(rightEye)
+
+            ear = (leftEAR + rightEAR) / 2.0
+        else:
+            continue
+
+        framewise.append([frame_n, ear])
+
+    vs.release()
```

Comment:
I just thought its standard practice to release the `VideoCapture` object using `vs.release()` once you are finished with capturing or processing the video to free up resources and video access. E.g. if someone runs analysis on a video, finishes and python remains open without the `vs.release`, I think they cant rename the video or move it, since it is still open on python and not released